research/object_detection/meta_architectures/context_rcnn_lib_tf2.py

```diff
@@ -0,0 +1,217 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Library functions for ContextRCNN."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+# The negative value used in padding the invalid weights.
+_NEGATIVE_PADDING_VALUE = -100000
+
+KEY_NAME = 'key'
+VALUE_NAME = 'val'
+QUERY_NAME = 'query'
+FEATURE_NAME = 'feature'
+
+class ContextProjection(tf.keras.layers.Layer):
+  """Custom layer to do batch normalization and projection."""
+  def __init__(self, projection_dimension, freeze_batchnorm, **kwargs):
+    self.batch_norm = tf.keras.layers.BatchNormalization(
+        epsilon=0.001,
+        center=True,
+        scale=True,
+        momentum=0.97,
+        trainable=(not freeze_batchnorm))
+    self.projection = tf.keras.layers.Dense(units=projection_dimension,
+                                            activation=tf.nn.relu6,
+                                            use_bias=True)
+    super(ContextProjection, self).__init__(**kwargs)
+
+  def build(self, input_shape):
+    self.batch_norm.build(input_shape)
+    self.projection.build(input_shape)
+
+  def call(self, input_features, is_training=False):
+    return self.projection(self.batch_norm(input_features, is_training))
+
+class AttentionBlock(tf.keras.layers.Layer):
+  """Custom layer to perform all attention."""
+  def __init__(self, bottleneck_dimension, attention_temperature,
+               freeze_batchnorm, output_dimension=None,
+               is_training=False, **kwargs):
+    self._key_proj = ContextProjection(bottleneck_dimension, freeze_batchnorm)
+    self._val_proj = ContextProjection(bottleneck_dimension, freeze_batchnorm)
+    self._query_proj = ContextProjection(bottleneck_dimension, freeze_batchnorm)
+    self._feature_proj = None
+    self._attention_temperature = attention_temperature
+    self._freeze_batchnorm = freeze_batchnorm
+    self._bottleneck_dimension = bottleneck_dimension
+    self._output_dimension = output_dimension
+    self._is_training = is_training
+    super(AttentionBlock, self).__init__(**kwargs)
+
+  def set_output_dimension(self, output_dim):
+    self._output_dimension = output_dim
+
+  def build(self, input_shapes):
+    pass
+
+  def call(self, box_features, context_features, valid_context_size):
+    """Handles a call by performing attention."""
+    _, context_size, _ = context_features.shape
+    valid_mask = compute_valid_mask(valid_context_size, context_size)
+    channels = box_features.shape[-1]
+
+    #Build the feature projection layer
+    if not self._output_dimension:
+      self._output_dimension = channels
+    if not self._feature_proj:
```

Comment:
Either way, after the first call it will be initialized. I think it would work the same whichever way it is done, but I think the way you suggest is more canonical so I'll go with that.